Cython/Utils.py

```diff
@@ -148,21 +148,31 @@ def find_root_package_dir(file_path):
 
 @cached_function
 def check_package_dir(dir, package_names):
+    namespace = True
     for dirname in package_names:
         dir = os.path.join(dir, dirname)
-        if not is_package_dir(dir):
-            return None
-    return dir
+        has_init = contains_init(dir)
+        if not namespace and not has_init:
+            return None, False
+        elif has_init:
+            namespace = False
+    return dir, namespace
 
 
 @cached_function
-def is_package_dir(dir_path):
+def contains_init(dir_path):
     for filename in PACKAGE_FILES:
         path = os.path.join(dir_path, filename)
         if path_exists(path):
             return 1
 
 
+@cached_function
```

Comment:
I don't think we need this decorator here. It's enough to have it on `contains_init()`.

Reply:
Removed decorator.